crates/red_knot/src/files.rs

```diff
@@ -0,0 +1,96 @@
+use std::fmt::{Debug, Formatter};
+use std::hash::{BuildHasherDefault, Hash, Hasher};
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+
+use hashbrown::hash_map::RawEntryMut;
+use parking_lot::RwLock;
+use rustc_hash::FxHasher;
+
+use ruff_index::{newtype_index, IndexVec};
+
+type Map<K, V> = hashbrown::HashMap<K, V, BuildHasherDefault<rustc_hash::FxHasher>>;
+
+#[newtype_index]
+pub struct FileId;
+
+// TODO we'll need a higher level virtual file system abstraction that allows testing if a file exists
+// or retrieving its content (ideally lazily and in a way that the memory can be retained later)
+// I suspect that we'll end up with a FileSystem trait and our own Path abstraction.
+#[derive(Clone, Default)]
+pub struct Files {
+    inner: Arc<RwLock<FilesInner>>,
+}
+
+impl Files {
+    pub fn intern(&self, path: &Path) -> FileId {
+        self.inner.write().intern(path)
+    }
+
+    pub fn path(&self, id: FileId) -> PathBuf {
+        // TODO this should return a unowned path
+        self.inner.read().path(id).to_path_buf()
+    }
+}
+
+impl Debug for Files {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let files = self.inner.read();
+        let mut debug = f.debug_map();
+        for item in files.iter() {
+            debug.entry(&item.0, &item.1);
+        }
+
+        debug.finish()
+    }
+}
+
+#[derive(Default)]
+struct FilesInner {
+    by_path: Map<FileId, ()>,
```

Comment:
So I found this a bit confusing, let me see if I understand it now. This looks like effectively a hashset of `FileId`, but the reason the name `by_path` makes sense for it is that we actually use the hash of the path as the hash of each `FileId` in the set, as a way to determine if a path is already in the set, without actually storing the path in the set. Is that about right?

Reply:
Yes, the idea is to avoid allocating each path twice: once in the `Map` and once in the index. 

It indeed is a `HashSet` but the `hashbrown::HashSet` doesn't provide access to `raw_entry` (at least the last time I checked). But it's not really a big deal because `HashSet<V> = HashMap<V, ()>` anyway.

Reply:
Right; using a hashmap with unit values as a hashset wasn't confusing to me; the confusing part is that it looks like a hashset of FileId but "really" it's a hashset of paths, with the file id as just kind of a placeholder.